Blogpost for KEP-4988 (#51582)
kubernetes/website: content/en/blog/_posts/XXXX-XX-XX-kubernetes-v1-34-snapshottable-api-server-cache.md

---
layout: blog
title: "Kubernetes v1.34: Snapshottable API server cache"
date: XXX
slug: kubernetes-v1-34-snapshottable-api-server-cache
author: >
  [Marek Siarkowicz](https://github.com/serathius) (Google)
draft: true
---

For years, the Kubernetes community has been on a mission to improve the stability and performance predictability of the API server.

> A major focus of this effort has been taming `LIST` requests, which have historically been a primary source of high memory usage and heavy load on the `etcd` datastore.

Reviewer: ```suggestion
A major focus of this effort has been taming **list** requests, which have historically been a primary source of high memory usage and heavy load on the `etcd` datastore.
```

With each release, we've chipped away at the problem, and today, we're thrilled to announce the final major piece of this puzzle.

> The **Snapshottable API Server Cache** feature is graduating to **Beta in Kubernetes v1.34**,

Reviewer: ```suggestion
The *snapshottable API server cache** feature has graduated to **Beta** in Kubernetes v1.34,
```

> The *snapshottable API server cache** feature has graduated to **Beta** in Kubernetes v1.34,

Reviewer: nit (might be my typo):
```suggestion
The *snapshottable API server cache* feature has graduated to **Beta** in Kubernetes v1.34,
```

culminating a multi-release effort to allow virtually all read requests to be served directly from the API server's cache.

> ## Evolving the Cache for Performance and Stability

Reviewer: nit: should be sentence case

The path to the current state involved several key enhancements over recent releases that paved the way for today's announcement.

> ### Consistent Reads from Cache (Beta in v1.31)

Reviewer: nit: should be sentence case

### Taming large responses with streaming (Beta in v1.33)

Another key improvement was tackling the problem of memory spikes when transmitting large responses. The streaming encoder, introduced in v1.33, allowed the API server to send list items one by one, rather than buffering the entire multi-gigabyte response in memory. This made the memory cost of sending a response predictable and minimal, regardless of its size.

### The missing piece

Despite these huge improvements, a critical gap remained. Any request for a historical `LIST`—most commonly used for paginating through large result sets—still had to bypass the cache and query `etcd` directly. This meant that the cost of *retrieving* the data was still unpredictable and could put significant memory pressure on the API server.

## Kubernetes 1.34: snapshots complete the picture

> The Snapshottable API Server Cache solves this final piece of the puzzle.

Reviewer: ```suggestion
The _snapshottable API server cache_ solves this final piece of the puzzle.
```

When a **list** request for a historical `resourceVersion` arrives, the API server now finds the corresponding snapshot and serves the response directly from its memory.
This closes the final major gap, allowing paginated requests to be served entirely from the cache.

## A new era of API Server performance 🚀

With this final piece in place, the synergy of these three features ushers in a new era of API server predictability and performance:

1. **Get Data from Cache**: *Consistent reads* and *snapshottable cache* work together to ensure nearly all read requests—whether for the latest data or a historical snapshot—are served from the API server's memory.
2. **Send data via stream**: *Streaming list responses* ensure that sending this data to the client has a minimal and constant memory footprint.

The result is a system where the resource cost of read operations is almost fully predictable and much more resiliant to spikes in request load.
This means dramatically reduced memory pressure, a lighter load on `etcd`, and a more stable, scalable, and reliable control plane for all Kubernetes clusters.

## How to get started

> With its graduation to Beta, the `SnapshottableCache` feature is **enabled by default** in Kubernetes v1.34. There are no actions required to start benefiting from these performance and stability improvements.

Reviewer: ```suggestion
With its graduation to Beta, the `SnapshottableCache` feature gate is **enabled by default** in Kubernetes v1.34. There are no actions required to start benefiting from these performance and stability improvements.
```

> With its graduation to Beta, the `SnapshottableCache` feature is **enabled by default** in Kubernetes v1.34. There are no actions required to start benefiting from these performance and stability improvements. The feature is controlled by the `ListFromCacheSnapshot` feature gate.

Author: It means you can disable the feature using `ListFromCacheSnapshot` feature gate.
